Math: Library: Added cordic sine function

ShriramShastry opened the pull request:
Implement cordic sine algorithm with angle in the range [-pi/2, pi/2] having total harmonic distortion (THD) of (-170) dBc for SNR

Signed-off-by: ShriramShastry <malladi.sastry@intel.com>

** figure -1 cordic sine performance** 

![image](https://user-images.githubusercontent.com/61411947/113000244-4214ae80-918d-11eb-8397-1d4dbd66cd76.png)

** figure -2 Computed time per sample derived using x86 performance count**

![image](https://user-images.githubusercontent.com/61411947/113249681-bc077d80-92dc-11eb-90cf-b546ca3db335.png)


** figure -3 Difference between floating point and fixed point out**
![image](https://user-images.githubusercontent.com/61411947/113562192-4cb6c400-9623-11eb-9eb2-3820bc22248e.png)

paulstelian97 commented:
I'd expect the commit message to contain a description on what this improves.

marc-hb commented:
Github Actions hang fixed by #3992 . @ShriramShastry you must:
```
git stash
git fetch <origin> pull/3990/head
git checkout FETCH_HEAD
git commit --amend # fake commit to change the date and SHA
git push <your remote> HEAD:refs/heads/shastry_cordicsin_dev
```
This will restart Github Actions. Sorry for the inconvenience.

Of course you could also address review comments and push, this would also trigger the same thing.

lgirdwood commented:
@ShriramShastry are you able to rebase on top of latest main branch to fix the CI report. We are good to merge, we just need to make sure your shastry_cordicsin_dev branch has the CI fix from the main branch.

lgirdwood commented:
@zrombel can you comment on UT here, as this should be unrelated ?

marc-hb commented:
In https://sof-ci.01.org/sof-pr-viewer/#/build/PR3990/build6249622 all platforms fail with something like this so it does seem related
```
39/39 Testing: sin_fixed
39/39 Test: sin_fixed
Command: "/localdisk/tools/xtensa/RG-2017.8-linux/XtensaTools/bin/xt-run" "--exit_with_target_code" "sin_fixed"
Directory: /quickbuild/workspace1/24733/SOF_FW/build_ut/test/cmocka/src/math/trig
"sin_fixed" start time: Apr 01 15:39 CEST
Output:
----------------------------------------------------------
1..1
test_math_trig_sin_fixed: diff for 1 deg = 0.0087261982
not ok 1 - test_math_trig_sin_fixed
# diff <= CMP_TOLERANCE
# /quickbuild/workspace1/24733/SOF_FW/test/cmocka/src/math/trig/sin_fixed.c:132: error: Failure!
# not ok - tests
<end of output>
Test time =   3.55 sec
----------------------------------------------------------
Test Failed.
"sin_fixed" end time: Apr 01 15:39 CEST
"sin_fixed" time elapsed: 00:00:03

97% tests passed, 1 tests failed out of 39

Total Test time (real) =  12.24 sec

The following tests FAILED:
	 39 - sin_fixed (Failed)
Errors while running CTest
```

lgirdwood commented:
@ShriramShastry can you make sure the UT failing above is updated with your new improved function. It could also be reporting a bug in the function.

marc-hb commented:
> Optimized cordicsine implementation with reduced lookup table with a faster, accurate and peformant

How faster? Just a rough order of magnitude.

ShriramShastry commented:
> > Optimized cordicsine implementation with reduced lookup table with a faster, accurate and peformant
> 


figure -1 
**_`Description - Comparative results between floating point and fixed point . Please see ERROR and total harmonic distortion parameters . Floating and fixed test measurements are done with Input range from [2*pi to 2*pi] values in step size of  pi/512.  `_**
figure -2
**_`Description - I did the measurement ( unroll the loop ) for each sample data ( 2048) , time ( in Nanoseconds) is captured on Y scale with inputs sample on X scale. Time measurement was done using x86 CPU  performance counter.`_**

> How faster? Just a rough order of magnitude.
**_`Improvement- Proposed implementation takes  approx. ¼ execution time with lookup table of [ int32 31 index ] with a saving of 481 int32 bytes when compared to GH`_**

ShriramShastry commented:
> @ShriramShastry can you make sure the UT failing above is updated with your new improved function. It could also be reporting a bug in the function.

Thank you @lgirdwood 
I haven't verified the input and scaling for floating point sine which is used to  compare the sine_fixed fcn , 
UT test failure is observed if floating to fixed result difference is greater than defined tolerance value of 0. 000005 then we hit error condition  
I will make some test and I will share my observations and findings

cujomalainey commented on src/math/trig.c:
```diff
@@ -9,568 +9,97 @@
+	/*cordic_atan2_lookup_table = atan(2.^-(0:N-1)) N = 31 */
+	/*CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:31-1)))) */
+	/*Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain */
+	static const int32_t inpLUT[31] = { 843314857, 497837829, 263043837, 133525159,
```
please give this a better name, inpLUT does not clearly state what it contains. A more appropriate name would be a truncated version fo the name you gave it above `atan2_lookup_table`

ShriramShastry replied:
OK, will replace it with a new name

cujomalainey commented on test/cmocka/src/math/trig/sin_fixed.c:
```diff
@@ -121,7 +121,7 @@ static void test_math_trig_sin_fixed(void **state)
 		int32_t rad_q28 = Q_CONVERT_FLOAT(rad, 28);
 
-		float r = Q_CONVERT_QTOF(sin_fixed(rad_q28), 31);
+		float r = Q_CONVERT_QTOF(sin_fixed(rad_q28), 30);
```
does this function replacement change the Q format returned?

ShriramShastry replied:
Yes, From Q1.31 to Q2.30. 
I missed to change it here since _int32_t sin_fixed(int32_t th_rad_fxp)_ is Q2.30 . As a result cmocka test failed . 
Please see figure -3 , plot displays difference between floating and fixed point sine for all 360 degree values

UT test will fail if floating to fixed result difference is greater than defined tolerance value of 0. 000005

cujomalainey replied:
hmm did we check to see if any users need to be adjusted who use this function. If yes we need to add additional commits to make them compatible

ShriramShastry replied:
Yes - this is needed
drc
_int32_t abs_sin_val = sin_fixed(q_mult(ABS(x), PI_OVER_TWO, 30, 30, 28));_
Tone generator
_sine = q_mults_32x32(sin_fixed(sg->w), sg->a, Q_SHIFT_BITS_64(31, 31, 31));_

ShriramShastry replied:
> Look pretty good, can you check the things mentioned by me and Keyon!

Added the changes as per required change.

> Look pretty good, can you check the things mentioned by me and Keyon!

Thanks a lot for review and inputs.

cujomalainey commented on src/audio/tone.c:
```diff
@@ -143,7 +143,7 @@ static int32_t tonegen(struct tone_state *sg)
 	int64_t w;
 
-	/* sg->w is angle in Q4.28 radians format, sin() returns Q1.31 */
+	/* sg->w is angle in Q2.30 radians format, sin() returns Q1.31 */
```
I think you changed the wrong thing here.

cujomalainey commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -113,7 +113,7 @@ inline int32_t drc_sin_fixed(int32_t x)
 	const int32_t PI_OVER_TWO = Q_CONVERT_FLOAT(1.57079632679489661923f, 30);
```
there is a comment above that is wrong, we need to check if any code that depends on this needs it math updated

cujomalainey commented on src/audio/tone.c:
```diff
@@ -74,7 +74,7 @@ struct tone_state {
 	int32_t fs; /* Sample rate in Hertz Q32.0 */
 	int32_t ramp_step; /* Amplitude ramp step Q1.31 */
-	int32_t w; /* Angle radians Q4.28 */
+	int32_t w; /* Angle radians Q2.30 */
```
shoulnt the angle be fine the way it is?

cujomalainey commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -113,7 +113,7 @@ inline int32_t drc_sin_fixed(int32_t x)
 
 	/* input range of sin_fixed() is non-negative */
-	int32_t abs_sin_val = sin_fixed(q_mult(ABS(x), PI_OVER_TWO, 30, 30, 28));
+	int32_t abs_sin_val = sin_fixed(q_mult(ABS(x), PI_OVER_TWO, 30, 30, 29));
```
i thought only the output was changed? why are we changing the input

johnylin76 replied:
Yeah it should not be changed.
And if you change the output q-format for sin_fixed(), you must take care of all existing callers to avoid errors.
I think you should add:
abs_sin_val = sat_int32(Q_SHIFT_LEFT((int64_t)abs_sin_val, 30, 31));

ShriramShastry replied:
OK - I will modify accordingly

aiChaoSONG commented:
SOFCI TEST

ShriramShastry commented:
> SOFCI TEST

Thanks @aiChaoSONG for helping me with triggering the tests.

lgirdwood commented:
@cujomalainey please merge if you are happy.

keyonjie commented on src/math/trig.c:
```diff
@@ -9,568 +9,97 @@
+	} else {
+		sign  = 1;
+	}
+	th_rad_fxp *= 4;
```
th_rad_fxp <<= 2;

ShriramShastry replied:
Yup, both have same result .

cujomalainey replied:
Yep but shift is faster :)

ShriramShastry replied:
I had a disassembly snippet, please see

```
th_rad_fxp   <<= 2;
00BB5881  mov         eax,dword ptr [th_rad_fxp  ]  
00BB5884  shl         eax,2  
00BB5887  mov         dword ptr [th_rad_fxp  ],eax  
th_rad_fxp   *= 4;
00BB588A  mov         eax,dword ptr [th_rad_fxp  ]  
00BB588D  shl         eax,2  
00BB5890  mov         dword ptr [th_rad_fxp  ],eax 

```
I see no advantage .  
But its fine with to me to change it

aiChaoSONG replied:
seems the compiler is smart enough to do an optimization like this, I try this in Xtensa Xplorer, even without `-O` optimization argument, the disassembly output is using shift instruction for both `num*4` and `num<<2`

keyonjie reviewed (commented):
Thanks for the optimization @ShriramShastry , do you intend to add support of input formats Q2.30, Q3.29 (besides today's Q4.28) for sin_fixed()? If so, better to add another param to denote this and avoid the checking logic. The variables declaration and comments need follow the guide and remained code also.

ShriramShastry commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -114,6 +114,7 @@ inline int32_t drc_sin_fixed(int32_t x)
 
 	/* input range of sin_fixed() is non-negative */
 	int32_t abs_sin_val = sin_fixed(q_mult(ABS(x), PI_OVER_TWO, 30, 30, 28));
+	abs_sin_val = sat_int32(Q_SHIFT_LEFT((int64_t)abs_sin_val, 30, 31));
```
function sin_fixed return value is Q2.30 and for compatible support ,   please see 
conversation link https://github.com/thesofproject/sof/pull/3990#discussion_r607344880

cujomalainey replied:
we should do this in the function itself though rather than hunting down all the places this is used, we can't fix inflight PRs that have not landed, so better to fix ourselves until we have a plan to fix the output.

ShriramShastry replied:
@cujomalainey , made the  change within the function itself i.e. from Q2.30 to Q1.30

ShriramShastry commented:
> Thanks for the optimization @ShriramShastry , do you intend to add support of input formats Q2.30, Q3.29 (besides today's Q4.28) for sin_fixed()? If so, better to add another param to denote this and avoid the checking logic. The variables declaration and comments need follow the guide and remained code also.


_do you intend to add support of input formats Q2.30, Q3.29 (besides today's Q4.28) for sin_fixed()_

The cordic sine algorithm only converges when the angle is in the range [-pi/2, pi/2). If an angle is outside of this range, then a multiple of pi/2 is added or subtracted from the angle until it is within the range [-pi/2,pi/2). The  cordicsin function do addition or subtraction by a multiple of pi/2 in the data type of the input. So, when the fraction length is 30 or 29 , then the quantization error introduced by the addition or subtraction of pi/2 is done on input fraction bits of precision.

I checked the implementation with cmocka which is pretty good https://github.com/thesofproject/sof/pull/3990#discussion_r607344880 
figure -3 in the draft explains

singalsu commented on src/math/trig.c:
```diff
@@ -9,568 +9,97 @@
+			xn -= ytmp;
+			b_yn += xtmp;
+		}
+	xtmp = xn >> (b_idx + 1);
```
Some code indentation issue here?

singalsu commented on src/math/trig.c:
```diff
@@ -9,568 +9,97 @@
-	delta = s1 - s0; /* Q1.31 */
-	/* All Q1.31 */
-	sine = s0 + q_mults_32x32(frac, delta, Q_SHIFT_BITS_64(31, 31, 31));
+		cdc_sin_th = sign * b_yn;
```
No need for if, just multiply with sign.

singalsu reviewed (commented):
Look pretty good, can you check the things mentioned by me and Keyon!

ShriramShastry commented on src/math/trig.c:
```diff
@@ -9,568 +9,97 @@
-	i1 = idx & (2 * SINE_NQUART - 1);
-	if (i1 > SINE_NQUART)
-		i1 = 2 * SINE_NQUART - i1;
+	if (th_rad_fxp > CORDICSINE_PIOVERTWO_Q28FL) {
```
Since the input range of cordicsin must be in the range [-2*pi, 2*pi), then we know that a signed type with fraction Length = wordLength-4 will fit this range without overflow.
In case you need additional param with a different input format, say Q8.24 ,then input will have less precision than Q4.28,
so upcasting to have the larger fraction Length does not change the value of the input, and acts like a left shift of the data. Thus increase of fraction Length makes the addition or subtraction of a multiple of pi/2 more precise.

cujomalainey replied:
Im with @ShriramShastry, i think it is up to the caller to make sure they are handing the correct bit format

ShriramShastry replied:
@cujomalainey , now we have cordicsine output back to  Q1.31 format.

paulstelian97 reviewed (commented):
Still cannot fully review algorithm (my own theory is lacking).

Marking this as comment rather than approve due to issues with the commit message itself (for example, you need to fix the range in the commit message -- should be `[-2*pi to 2*pi]` -- and don't insert spaces where you shouldn't -- should be `sizeof(int32)` without extra spaces between the parentheses).

Other than that, as far as I could tell this looks good. I assume tests confirm that this works correctly.

cujomalainey commented on src/math/trig.c:
```diff
@@ -9,568 +9,115 @@
+	 *CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:31-1))))
+	 *Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain
+	 */
+	static const int32_t cordicsine_lookup[31] = { 843314857, 497837829, 263043837, 133525159,
```
i just realized this is `1 << idx` can we update this to use the `BIT` macro? E.g.
```
arr = {
  BIT(31),
  BIT(30),
  ...
  BIT(1),
  BIT(0)
};
```

ShriramShastry replied:
This is not a bit table (2^N) instead its a lookuptable  having 0 to 31 bits. There is a description/comment available in the code with formula above variable cordicsine_lookup[31].

cujomalainey replied:
got it thanks

cujomalainey commented on src/math/trig.c:
```diff
@@ -9,568 +9,115 @@
+	xn = CORDICSINE_LUT_Q29FL;
+	xtmp = CORDICSINE_LUT_Q29FL;
+	ytmp = 0;
+	for (b_idx = 0; b_idx < 31; b_idx++) {
```
should make a comment here about why this is 31 and not 32

ShriramShastry replied:
It's iterating over lookup table index

cujomalainey replied:
exactly, it looks like you are using 32bits, you should use ARRAY_SIZE then

ShriramShastry replied:
Sure , thanks

cujomalainey commented on src/math/trig.c:
```diff
@@ -9,568 +9,115 @@
+	else if (th_rad_fxp <= CORDICSINE_N57DEG_Q30FL)
+		th_rad_fxp = INT32_MIN;
+	else
+		th_rad_fxp <<= 1;
```
left shift on fixed point, don't we have a macro so cppcheck wont get mad?

ShriramShastry replied:
th_rad_fxp <<= 1 is  th_rad_fxp *= 2.

cujomalainey replied:
left shift is faster, but we have a macro because left shift on signed values are technically invalid if i remember correctly

ShriramShastry replied:
OK - then I add left shift macro

aiChaoSONG replied:
@cujomalainey left shift on signed values is valid, but if you got a positive number by shift a negative number or got a negative value by shift a positive number, you may experience overflow or downflow. I think the macro is used to take case of overflow or downflow.

cujomalainey commented on src/math/trig.c:
```diff
@@ -9,568 +9,115 @@
+	int32_t xn;
+	int32_t xtmp;
+	int32_t ytmp;
+	int32_t cdc_sin_th;
```
this can be deleted

ShriramShastry replied:
Yep, this can be done. 
Is done to keep input and return value name different

cujomalainey commented on src/math/trig.c:
```diff
@@ -9,568 +9,115 @@
-	/* All Q1.31 */
-	sine = s0 + q_mults_32x32(frac, delta, Q_SHIFT_BITS_64(31, 31, 31));
+	cdc_sin_th = th_rad_fxp;
+	return cdc_sin_th;
```
why not just return th_rad_fxp?

ShriramShastry replied:
Yep, this can be done.
Is done to keep input and return value name different

cujomalainey replied:
i dont think it matters, adding extra values looks like a bug.

ShriramShastry replied:
Fine , then I return th_rad_fxp instead of cdc_sin_th

ShriramShastry commented:
> Still cannot fully review algorithm (my own theory is lacking).
> 
> Marking this as comment rather than approve due to issues with the commit message itself (for example, you need to fix the range in the commit message -- should be `[-2*pi to 2*pi]` -- and don't insert spaces where you shouldn't -- should be `sizeof(int32)` without extra spaces between the parentheses).
> 
> Other than that, as far as I could tell this looks good. I assume tests confirm that this works correctly.

Thanks you. 
I will include your suggested change in my next commit message.
A description of cordicsine is available  in https://en.wikipedia.org/wiki/CORDIC)

singalsu commented on src/math/trig.c:
```diff
@@ -9,568 +9,95 @@
+			sign  = -1;
+		} else {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
```
Can this `sign = 1; `set be left out?

ShriramShastry replied:
Yes. this can be done as it is set to 1 in the beginning.

singalsu commented:
@ShriramShastry Also please mark as resolved the issues those you have addressed.

ShriramShastry commented on src/math/trig.c:
```diff
@@ -9,568 +9,95 @@
+			sign  = -1;
+		} else {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
```
sign = 1;
Is this also should be left out ?

singalsu reviewed (approved):
Looks good to me, thanks!

lgirdwood commented:
CML showing unrelated kernel resume issue.
